## HW 7 - Hongyi Tang

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.formula.api as smf

## 1 Heart Attack Prediction

1.1

In [23]:
heart = pd.read_csv("../data/heart.csv")
heart.head(3)

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1

In [24]:
# drop unwanted columns
heart.drop(["slp", "oldpeak", "thall"], axis=1, inplace=True)

In [25]:
# check the shape of heart
heart.shape

(303, 11)

In [26]:
# check NAs in this csv file
heart.isna().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
caa         0
output      0
dtype: int64

In [27]:
# desriptive statistics
heart.describe()

age         sex          cp      trtbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg    thalachh        exng         caa      output  
count  303.000000  303.000000  303.000000  303.000000  303.000000  
mean     0.528053  149.646865    0.326733    0.729373    0.544554  
std      0.525860   22.905161    0.469794    1.022606    0.498835  
min      0.000000   71.000000    0.000000    0.000000    0.000000  
25%      0.000000  133.500000    0.000000    0.000000    0.000000  
50%      1.000000  153.000000    0.000000    0.000000    1.000000  
75%      1.000000  166.000000    1.000000    1.000000    1.000000  
max      2.000000  202.000000    1.000000    4.000000    1.000000

1.2

In [28]:
# transform to categorical variable
m = smf.logit('output ~ age+sex+C(cp)+trtbps+chol+fbs+C(restecg)+thalachh+exng+C(caa)', data=heart).fit()

Optimization terminated successfully.
         Current function value: 0.355396
         Iterations 7


In [29]:
m.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 output   No. Observations:                  303
Model:                          Logit   Df Residuals:                      286
Method:                           MLE   Df Model:                           16
Date:                Mon, 07 Mar 2022   Pseudo R-squ.:                  0.4843
Time:                        17:23:29   Log-Likelihood:                -107.69
converged:                       True   LL-Null:                       -208.82
Covariance Type:            nonrobust   LLR p-value:                 2.759e-34
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           0.4035      2.469      0.163      0.870      -4.435       5.242
C(cp)[T.1]          1.5343      0.543      2.827      0.005       0.470       2.598
C(cp)[T.2]          1.6907      0.441      3.838      0.000       0.827       2.554
C(cp)[T.3]          1.8970      0.667      2.843      0.004       0.589       3.205
C(restecg)[T.1]     0.4206      0.361      1.165      0.244      -0.287       1.128
C(restecg)[T.2]    -0.9518      2.118     -0.449      0.653      -5.103       3.199
C(caa)[T.1]        -1.8571      0.458     -4.052      0.000      -2.756      -0.959
C(caa)[T.2]        -3.0485      0.647     -4.712      0.000      -4.317      -1.780
C(caa)[T.3]        -2.3741      0.783     -3.032      0.002      -3.908      -0.840
C(caa)[T.4]         0.4676      1.586      0.295      0.768      -2.642       3.577
age                 0.0237      0.024      0.992      0.321      -0.023       0.070
sex                -2.1367      0.457     -4.671      0.000      -3.033      -1.240
trtbps             -0.0285      0.011     -2.656      0.008      -0.050      -0.007
chol               -0.0061      0.004     -1.524      0.128      -0.014       0.002
fbs                 0.4437      0.501      0.886      0.375      -0.537       1.425
thalachh            0.0357      0.011      3.391      0.001       0.015       0.056
exng               -1.1615      0.418     -2.781      0.005      -1.980      -0.343
===================================================================================
"""

In [30]:
m.get_margeff().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
        Logit Marginal Effects       
=====================================
Dep. Variable:                 output
Method:                          dydx
At:                           overall
===================================================================================
                     dy/dx    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
C(cp)[T.1]          0.1716      0.058      2.975      0.003       0.059       0.285
C(cp)[T.2]          0.1891      0.045      4.234      0.000       0.102       0.277
C(cp)[T.3]          0.2122      0.071      2.986      0.003       0.073       0.351
C(restecg)[T.1]     0.0470      0.040      1.173      0.241      -0.032       0.126
C(restecg)[T.2]    -0.1065      0.237     -0.450      0.653      -0.570       0.358
C(caa)[T.1]        -0.2077      0.046     -4.532      0.000      -0.298      -0.118
C(caa)[T.2]        -0.3410      0.062     -5.515      0.000      -0.462      -0.220
C(caa)[T.3]        -0.2655      0.082     -3.241      0.001      -0.426      -0.105
C(caa)[T.4]         0.0523      0.177      0.295      0.768      -0.295       0.400
age                 0.0026      0.003      0.998      0.318      -0.003       0.008
sex                -0.2390      0.045     -5.331      0.000      -0.327      -0.151
trtbps             -0.0032      0.001     -2.776      0.006      -0.005      -0.001
chol               -0.0007      0.000     -1.539      0.124      -0.002       0.000
fbs                 0.0496      0.056      0.890      0.374      -0.060       0.159
thalachh            0.0040      0.001      3.628      0.000       0.002       0.006
exng               -0.1299      0.045     -2.912      0.004      -0.217      -0.042
===================================================================================
"""

(a): males have 23.90 pct points larger chance to get heart attack, compared to females, if their other characteristics are the same. It is statistically significant since the p-value is 0

(b): comparing with patient who has typical angina chest pain, patient who has atypical angina have 17.16 pct point higher chance of getting the heart attack;this is statistically significant since the p value is very low (0.003). Just like that, patient who has non-anginal pain has 18.91 pct point higher chance of getting the heart attack; this is statistically significant since the p value is very low (0). Patient who has asymptomatic pain has 21.22 pct point higher chance of getting a heart attack; it is also statistically significant since the p value is very low (0.003) 

(c): when there is 1 unit increase on age, the probability of getting heart attack increase 0.26 pct point. it is not staticsally significant since the p value is very large (0.318)

*(d): 
- SS. stands for statistical significance
     
| Variable    | SS.         |Intuitive|
| ----------- | ----------- |----------|
| restecg - Value2   | No        |No|
| chol   | No        |No|
| exng   | Yes        |No|
| caa - 0   | Yes        |No|
| caa - 1   | Yes        |No|
| caa - 2   | Yes        |No|
| sex   | Yes        |No|
| trtbps   | Yes        |No|

Why:
restecg - Value2 - It doesn't make sense that people who have ST-T wave abnormality, the probability of gettiing heart attack is decreasing

Sex - It doesn't make sense that one gender is more likely to get heart attack than the other

trtbps - It doesn't make sense that someone with higher blood pressure has a lower chance of getting a heart attack

chol - It doesn't make sense that people with higher cholesterol level has a lower chance of getting a heart attack

exng - It doesn't make sense that exercise induced angina will lower the chance of gettint a heart attack

caa(0,1,2,3) - I believe that it is not intuitive since more major vessels should better preventing heart attacks, and people with few vessels should be more likely to get heart attacks

1.3

In [31]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [32]:
var = heart.drop(['output'], axis=1)
X = pd.get_dummies(var, drop_first = True, columns = ['cp', 'restecg', 'caa'])
y = heart['output']

In [33]:
logistic = LogisticRegression(penalty='none', solver='newton-cg').fit(X,y)

In [34]:
# coefficient
logistic.coef_

array([[ 0.02366493, -2.13671462, -0.02850622, -0.00606629,  0.44373721,
         0.03574632, -1.16147349,  1.534265  ,  1.69073286,  1.89700888,
         0.42062829, -0.9517973 , -1.8571451 , -3.0484795 , -2.37405273,
         0.46760006]])

In [35]:
# intercept
logistic.intercept_

array([0.4034782])

I get the same intercept and coefficients as what I got from the statsmodel package

1.4

1.41

In [36]:
# probability of having heart attacks 
train_pred = logistic.predict_proba(X)
train_pred[:10]

array([[0.12858798, 0.87141202],
       [0.06204112, 0.93795888],
       [0.01609467, 0.98390533],
       [0.0449013 , 0.9550987 ],
       [0.21926213, 0.78073787],
       [0.4573165 , 0.5426835 ],
       [0.04936884, 0.95063116],
       [0.08085049, 0.91914951],
       [0.15019385, 0.84980615],
       [0.06587713, 0.93412287]])

1. first column: $P(output=0|X)$
2. second column: $P(output=1|X)$

In [37]:
# take out the second column, which are the probabilities of output=1.
train_predict = train_pred[:,1]
# select the first 10 probabilities 
train_predict[:10]

array([0.87141202, 0.93795888, 0.98390533, 0.9550987 , 0.78073787,
       0.5426835 , 0.95063116, 0.91914951, 0.84980615, 0.93412287])

1.42

In [38]:
## directly predict the labels for wherher or not the person will have a heart attack: either 1 or 0.
labels = logistic.predict(X)
labels[:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

1.43

In [41]:
# set up a threshold
threshold = 0.5

# compare probabilities of y=1 with the threshold:
# if probability > threshold, predict y=1
# if probability <= threshold, predict y=0

1.0*(train_predict>threshold)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.,
       0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1.,
       0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 0., 1., 0., 0., 1., 0., 1.,
       1., 1., 1., 1., 0., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1.,
       1., 1., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0.,
       0., 1., 1., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1., 1., 0., 0., 0.,
       0., 1., 1., 1., 0., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.,
       0., 1., 0., 0., 0.

1.5

In [42]:
# accuracy socre
accuracy_score(labels,y)

0.858085808580858

About 85% of the predicted labels match with the true labels on the this training set. I'm comfortable enough to depoly this model to the real life since the accuracy score 0.85 is pretty high for predictions

1.6

In [43]:
# create a confusion matrix
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score

In [44]:
yhat = logistic.predict(X)
cm = confusion_matrix(y, yhat)
cm

array([[113,  25],
       [ 18, 147]])

In [45]:
# accuracy_score
accuracy_score(y, yhat)

0.858085808580858

In [46]:
# precision_score
precision_score(y, yhat)

0.8546511627906976

In [47]:
# recall_score
recall_score(y, yhat)

0.8909090909090909

## 2  Airbnb Price Prediction

**2.1**

In [48]:
airbnb = pd.read_csv("../data/airbnb-beijing-listings.csv.bz2", sep=",", \
                     usecols=['price', 'bedrooms', 'room_type', 'accommodates', 'bathrooms','host_is_superhost'])
airbnb.head(3)

host_is_superhost        room_type  accommodates  bathrooms  bedrooms  \
0                 f  Entire home/apt             9        2.0       3.0   
1                 f     Private room             2        1.0       1.0   
2                 t  Entire home/apt             2        1.0       1.0   

       price  
0    $831.00  
1  $1,197.00  
2    $598.00

In [49]:
airbnb['price'] = airbnb['price'].replace({'\$':'' , ',':''}, regex=True)
airbnb.price = airbnb.price.astype(float)

In [50]:
airbnb.isna().sum()

host_is_superhost     72
room_type              0
accommodates           0
bathrooms              7
bedrooms             114
price                  0
dtype: int64

In [51]:
airbnb = airbnb.dropna()

In [52]:
airbnb.isna().sum()

host_is_superhost    0
room_type            0
accommodates         0
bathrooms            0
bedrooms             0
price                0
dtype: int64

In [53]:
# filter out airbnbs where price is 0 
airbnb_cleaned = airbnb[airbnb['price'] !=0]
airbnb_cleaned.shape

(38614, 6)

In [54]:
airbnb_cleaned_copy = airbnb_cleaned.copy()
airbnb_cleaned_copy['bedroom_catgory'] = pd.cut(airbnb_cleaned_copy.bedrooms, 
                                               bins = [-1, 0, 1, 2, 3, 100],
                                               labels= ['0', '1', '2', '3', '4+'])
airbnb_cleaned_copy.head(3)

host_is_superhost        room_type  accommodates  bathrooms  bedrooms  \
0                 f  Entire home/apt             9        2.0       3.0   
1                 f     Private room             2        1.0       1.0   
2                 t  Entire home/apt             2        1.0       1.0   

    price bedroom_catgory  
0   831.0               3  
1  1197.0               1  
2   598.0               1

In [55]:
airbnb_ul = airbnb_cleaned_copy.copy()
airbnb_ul['accommodates'] = pd.cut(airbnb_ul.accommodates, 
                                               bins = [ 0, 1, 2, 3, 1000],
                                               labels= ['1', '2', '3', '4+'])

In [56]:
airbnb_ul['bathrooms'] = pd.cut(airbnb_ul.bathrooms, 
                                               bins = [ -1, 0, 1, 2, 100],
                                               labels= ["0",'1', '2', '3+'])
airbnb_ul.head(5)

host_is_superhost        room_type accommodates bathrooms  bedrooms   price  \
0                 f  Entire home/apt           4+         2       3.0   831.0   
1                 f     Private room            2         1       1.0  1197.0   
2                 t  Entire home/apt            2         1       1.0   598.0   
3                 t  Entire home/apt            2         1       1.0   598.0   
4                 t  Entire home/apt            2         1       1.0   598.0   

  bedroom_catgory  
0               3  
1               1  
2               1  
3               1  
4               1

In [57]:
airbnb_ul['log_price'] = np.log(airbnb_ul['price'])
airbnb_ul.head(3)

host_is_superhost        room_type accommodates bathrooms  bedrooms   price  \
0                 f  Entire home/apt           4+         2       3.0   831.0   
1                 f     Private room            2         1       1.0  1197.0   
2                 t  Entire home/apt            2         1       1.0   598.0   

  bedroom_catgory  log_price  
0               3   6.722630  
1               1   7.087574  
2               1   6.393591

**2.2**

In [58]:
# Replicate the model from PS6
reg2 = smf.ols("log_price ~ bedroom_catgory + accommodates + room_type + bathrooms", data=airbnb_ul).fit()
reg2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              log_price   R-squared:                       0.458
Model:                            OLS   Adj. R-squared:                  0.458
Method:                 Least Squares   F-statistic:                     2723.
Date:                Mon, 07 Mar 2022   Prob (F-statistic):               0.00
Time:                        17:24:11   Log-Likelihood:                -35615.
No. Observations:               38613   AIC:                         7.126e+04
Df Residuals:                   38600   BIC:                         7.137e+04
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
=============================================================================================
                                coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------
Intercept                     5.5801      0.062     89.815      0.000       5.458       5.702
bedroom_catgory[T.1]          0.0553      0.040      1.386      0.166      -0.023       0.134
bedroom_catgory[T.2]          0.1900      0.041      4.644      0.000       0.110       0.270
bedroom_catgory[T.3]          0.4917      0.042     11.632      0.000       0.409       0.575
bedroom_catgory[T.4+]         0.8731      0.044     19.945      0.000       0.787       0.959
accommodates[T.2]             0.3321      0.013     24.638      0.000       0.306       0.359
accommodates[T.3]             0.3935      0.017     23.646      0.000       0.361       0.426
accommodates[T.4+]            0.6087      0.015     39.342      0.000       0.578       0.639
room_type[T.Private room]    -0.3243      0.007    -43.874      0.000      -0.339      -0.310
room_type[T.Shared room]     -0.9507      0.017    -56.418      0.000      -0.984      -0.918
bathrooms[T.1]             2.267e-05      0.047      0.000      1.000      -0.092       0.092
bathrooms[T.2]                0.0343      0.047      0.722      0.470      -0.059       0.127
bathrooms[T.3+]               0.5907      0.050     11.883      0.000       0.493       0.688
==============================================================================
Omnibus:                     8914.379   Durbin-Watson:                   1.781
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            51799.993
Skew:                           0.986   Prob(JB):                         0.00
Kurtosis:                       8.320   Cond. No.                         53.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**2.3**

In [59]:
# predict the price of these listings from above model
phat = reg2.predict(airbnb_ul)
phat.head(3)

0    6.714877
1    5.643309
2    5.967570
dtype: float64

In [60]:
# calculate the RMSE 
np.sqrt(np.mean((phat - airbnb_ul["log_price"])**2))

0.6086074931250635

**2.4**

In [61]:
two_bedrooms = airbnb_ul[airbnb_ul['bedrooms']==2.0]
two_b_four = two_bedrooms[two_bedrooms['accommodates'].isin(["4+"])]
two_b_four

host_is_superhost        room_type accommodates bathrooms  bedrooms  \
9                     f  Entire home/apt           4+         2       2.0   
13                    t  Entire home/apt           4+         1       2.0   
36                    t  Entire home/apt           4+         1       2.0   
51                    f  Entire home/apt           4+         1       2.0   
65                    f  Entire home/apt           4+         1       2.0   
...                 ...              ...          ...       ...       ...   
38799                 f  Entire home/apt           4+         1       2.0   
38800                 f  Entire home/apt           4+         1       2.0   
38803                 f  Entire home/apt           4+         2       2.0   
38804                 f  Entire home/apt           4+         1       2.0   
38807                 f  Entire home/apt           4+         2       2.0   

        price bedroom_catgory  log_price  
9       669.0               2   6.505784  
13      950.0               2   6.856462  
36      704.0               2   6.556778  
51      690.0               2   6.536692  
65      289.0               2   5.666427  
...       ...             ...        ...  
38799   401.0               2   5.993961  
38800   591.0               2   6.381816  
38803  1000.0               2   6.907755  
38804   317.0               2   5.758902  
38807   500.0               2   6.214608  

[6881 rows x 8 columns]

In [62]:
# my prediction
predict = reg2.predict(two_b_four)
predict

9        6.413156
13       6.378887
36       6.378887
51       6.378887
65       6.378887
           ...   
38799    6.378887
38800    6.378887
38803    6.413156
38804    6.378887
38807    6.413156
Length: 6881, dtype: float64

**2.5**

In [63]:
# avg price for 2BR that acc 4 
avg_price = two_b_four.log_price.mean()
avg_price

6.385029736532868

In [64]:
# the avg of my prediction
predict_mean = predict.mean()
predict_mean

6.376677399984822

In [65]:
diff = avg_price - predict_mean
diff

0.008352336548045791

I got it pretty close. the difference is only 0.008352336548045791 between the actual average and the prediction average. The actual average is 0.008352336548045791 higher than my prediction average

## Total Time: 20hrs